Lib/os.py

```diff
@@ -343,86 +343,96 @@ def walk(top, topdown=True, onerror=None, followlinks=False):
     return _walk(fspath(top), topdown, onerror, followlinks)
 
 def _walk(top, topdown, onerror, followlinks):
-    dirs = []
-    nondirs = []
-    walk_dirs = []
-
-    # We may not have read permission for top, in which case we can't
-    # get a list of the files the directory contains.  os.walk
-    # always suppressed the exception then, rather than blow up for a
-    # minor reason when (say) a thousand readable directories are still
-    # left to visit.  That logic is copied here.
-    try:
-        # Note that scandir is global in this module due
-        # to earlier import-*.
-        scandir_it = scandir(top)
-    except OSError as error:
-        if onerror is not None:
-            onerror(error)
-        return
+    stack = [(False, top)]
```

Comment:
Will it make more sense to use a deque here? They are optimized for appending and popping while having the same interface.

Reply:
I think when append and pop occur only at the end (as in this case), performance of list and deque is basically the same. When append/pop can occur at the start, then you need deque ("double-ended queue") to avoid O(n) copy.

Reply:
In theory I think a deque is a slightly better fit, since we're only popping and appending (at a certain point `list` has to reallocate right?), but I don't think there's much of a difference in this case. I changed it to a deque, but can change back to a list if that's preferred

Reply:
Yes, list has to `realloc` occasionally, but proportional over-allocation means that this is amortized linear cost over a long sequence of appends (as in this use case). Deque in the same usage scenario has to repeatedly allocate new "chunks" of a fixed size, so this is also linear cost. I wrote a small microbenchmark modeling "hundreds of thousands of appends, then pop them all off" and there is no detectable performance difference on my machine. I personally don't think deque has any advantage here that makes it worth the import (which has to be inline to avoid a module-level dependency, adding some cost to every call.)

Reply:
Yeah, good points. I just did some rough benchmarking as well and found no real difference.

Also, I didn't find the exact cause, but adding `from collections import deque` to the top of `Lib/os.py` seemed to cause test failures under `test_embed` and `test_site`. Moving the import inside of the `walk` function seems to work, but at that point I agree it's not worth it.

I'll switch back to list